src/Spatial/Euclidean/StraightLine2D.cs

```diff
@@ -0,0 +1,312 @@
+        {
+            Gradient = gradient;
+            if (double.IsInfinity(gradient))
+            {
+                this.yoffset = double.NaN; // special case vertical line has no y-intercept
+                this.xoffset = x;
+            }
+            else
+            {
+                this.yoffset = y - (gradient * x);
+                this.xoffset = double.NaN;
+            }
+        }
+
+        /// <summary>
+        /// Determines if the line is horizontal
+        /// </summary>
+        /// <returns>True if the line is horizontal</returns>
+        public bool IsHorizontal => Gradient == 0;
+
+        /// <summary>
+        /// Determines if the line is vertical
+        /// </summary>
+        /// <returns>returns true if the line is vertical</returns>
+        public bool IsVertical => double.IsInfinity(Gradient);
+
+        /// <summary>
+        /// Tests if the other line is parallel to this one
+        /// </summary>
+        /// <returns>Returns true if lines are parallel</returns>
+        public bool IsParallel<T>(T otherLine) where T : ILine2D
+        {
+            if (otherLine is StraightLine2D)
+            {
+                IsParallel(otherLine);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Tests if the other line is parallel to this one
+        /// </summary>
+        /// <returns>Returns true if lines are parallel</returns>
+        public bool IsParallel(StraightLine2D otherLine)
+        {
+            return otherLine.Gradient == Gradient;
+        }
+
+        /// <summary>
+        /// Finds the midpoint of two points along the current line
+        /// </summary>
+        /// <param name="pointA">a point on the line</param>
+        /// <param name="pointB">a point on the line</param>
+        /// <returns>A point midway between the two provided points</returns>
+        public Point2D MidPoint(Point2D pointA, Point2D pointB)
+        {
+            if (IsThrough(pointA) && IsThrough(pointB))
+            {
+                return new Point2D((pointA.X + pointB.X) / 2, (pointA.Y + pointB.Y) / 2);
+            }
+
+            throw new ArgumentException("Provided points are not along the line");
+        }
+
+        /// <summary>
+        /// Returns a line perpendicular to this line going through point p
+        /// </summary>
+        /// <param name="p">The point on the line which it passes through</param>
+        /// <returns>A perpendicular straightline</returns>
+        public StraightLine2D PerpendicularAt(Point2D p)
+        {
+            return StraightLine2D.Create(p, -1 / Gradient);
+        }
+
+        /// <summary>
+        /// Find the intersection between two lines
+        /// </summary>
+        /// <param name="line">The line to intersect with</param>
+        /// <returns>The point of intersection or null if the lines are parallel</returns>
+        public Point2D? Intersection(StraightLine2D line)
```

Comment:
I think we should deprecate the Intersection that returns null in other parts of the lib and make them `public bool TryIntersectWith(StraightLine2D line, out Point2D result)` what do you think about this?

Reply:
not a huge fan to be honest, breaks the flow of using it